keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+for credential reverification (and therefore ignored for caching).
+
+The rules that define what should be ignored are set by the `imagePullCredentialsVerificationPolicy`:
+- `NeverVerify` matches the previous behavior from user perspective. Kubelet still creates
+   records of image pulls as these would be required should the policy change later.
+- `NeverVerifyPreloadedImages` images pulled without a secret and those pulled
+  outside the kubelet process are ignored (new default behavior)
+- `NeverVerifyAllowlistedImages` - images pulled without a secret and those pulled
+  outside the kubelet process but mentioned in `preloadedImagesVerificationAllowlist` are ignored
+- `AlwaysVerify` - all the image pulls need reverification.
+
+#### Writing to the Cache
+
+There are two different stages when the kubelet updates its cache:
+1. **Before the image pull**
+    - An image pull intent record is written before the actual pull occurs.
+    - This is is to mitigate a situation where the kubelet crashes during the pull and an
+      image that gets pulled in the meantime later appears as if it were *preloaded*.
+2. **After the pull**
+    - Once we know the result of the pull operation, we either remove the image pull intent record
+      (=> failed pull attempt), or we write a new "image pulled record" about the image
+      being pulled successfully and remove the original record of the image pull intent.
+    - Without this, if kubelet failed during an image pull, an existing image pull intent record
+      would already grant image access irregardless of this pull operation result for any next
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image and should be called
+    // before a pull of the image occurs.
+    //
+    // RecordPullIntent() should be called before every image pull. Each call of
+    // RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordPullIntent(image string) error
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
```

Comment:
I don't think we'd expect to call this to deal with dangling intent records recovering from a crash... that would happen internally on construction.

I think every call to RecordImagePulled *must* provide a non-nil `credentials *imagemanager.ImagePullCredentials`, with *either* a single image pull secret OR NodePodsAccessible=true